sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeClientOptions.cs

```diff
@@ -1,19 +1,222 @@
+            /// </summary>
+            V2019_02_02 = 1
+#pragma warning restore CA1707 // Identifiers should not contain underscores
+        }
+
+        /// <summary>
+        /// Gets the <see cref="ServiceVersion"/> of the service API used when
+        /// making requests.  For more, see
+        /// <see href="https://docs.microsoft.com/en-us/rest/api/storageservices/versioning-for-the-azure-storage-services" />.
+        /// </summary>
+        public ServiceVersion Version { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataLakeClientOptions"/>
+        /// class.
+        /// </summary>
+        /// <param name="version">
+        /// The <see cref="ServiceVersion"/> of the service API used when
+        /// making requests.
+        /// </param>
+        public DataLakeClientOptions(ServiceVersion version = LatestVersion)
+        {
+            Version = version == ServiceVersion.V2019_02_02 ? version : throw Errors.VersionNotSupported(nameof(version));
+            this.Initialize();
+            AddHeadersAndQueryParameters();
+        }
+
+        /// <summary>
+        /// Gets or sets the secondary storage <see cref="Uri"/> that can be read from for the storage account if the
+        /// account is enabled for RA-GRS.
+        ///
+        /// If this property is set, the secondary Uri will be used for GET or HEAD requests during retries.
+        /// If the status of the response from the secondary Uri is a 404, then subsequent retries for
+        /// the request will not use the secondary Uri again, as this indicates that the resource
+        /// may not have propagated there yet. Otherwise, subsequent retries will alternate back and forth
+        /// between primary and secondary Uri.
+        /// </summary>
+        public Uri GeoRedundantSecondaryUri { get; set; }
+
+        /// <summary>
+        /// Add headers and query parameters in <see cref="DiagnosticsOptions.LoggedHeaderNames"/> and <see cref="DiagnosticsOptions.LoggedQueryParameters"/>
+        /// </summary>
+        private void AddHeadersAndQueryParameters()
+        {
+            Diagnostics.LoggedHeaderNames.Add("x-ms-properties");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-client-request-id");
+            Diagnostics.LoggedHeaderNames.Add("If-Modified-Since");
+            Diagnostics.LoggedHeaderNames.Add("If-Unmodified-Since");
+            Diagnostics.LoggedHeaderNames.Add("If-Match");
+            Diagnostics.LoggedHeaderNames.Add("If-None-Match");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-rename-source");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-source-lease-id");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-permissions");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-umask");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-owner");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-group");
+            Diagnostics.LoggedHeaderNames.Add("x-ms-acl");
```

Comment:
Are these specific to Datalake?

Do the ones below need to be kept in sync with Blobs? If so, should we pull out into StorageClientOptions?

Reply:
We don't want to pull them into StorageClientOptions because Files and Queues have their own different sets.

Reply:
Yes, it would need to be just a method name convention or something like AddBlobDatalakeHeaders, which isn't great.

Reply:
But at least we wouldn't need to worry about them getting out of sync.

Reply:
If you want to do this, just make the method in BlobClientOptions `internal static` since we've already got visibility if we're calling the internal ctor.

Reply:
Yes, those are specific to Data Lake.